Add hints.selectors setting

olmokramer opened the pull request:
Make it possible for users to configure the CSS selectors that are used to determine which elements on the page should be hinted with the `hints.selectors` setting. The setting supports patterns, i.e. is configurable on a per-domain basis. As discussed below with @jgkamat I have removed `webelem.SELECTORS`, the default selectors are now defined in `configdata.yml`, so users can also remove selectors on specific pages if they want to. It is also possible to add additional groups to the setting dict, that can be used with the `:hint` command, for example:

```python
config.set('hints.selectors', {
    'expando': ['.expando-button'],
}, '*://*.reddit.com/*')
```

makes the expando buttons on reddit hintable with `:hint expando`. When a dict is set for the current domain (like with the command above) but the riekwested group isn't configured for this domain, the implementation first falls back to the global setting before it throws an error that the group doesn't exist.

TODO:

- [ ] Tests:
  * [x] Fix tests that relied on `webelem.SELECTORS` or `webelem.Group`.
  * [x] Add tests for the new behavior. Didn't really know where to start though.
    - [x] Element with a JS click handler
    - [x] Error on invalid hint group
    - [x] Custom hint groups
    - [x] Domain specific hints. How would I go about this? I could add an HTML file with a `<div class="clickable">` and a JS click handler on that div, but what domain pattern can I use?
    - [x] Fallback to global value when domain specific hints does not contain requested group.
    - [ ] More?
- [ ] Validate user input: check if the user selectors are actually valid. Right now the value of `hints.selectors` must be a dict of lists of strings, but those strings may be invalid. Couldn't find anything sensible in the list of config types, so maybe a `CSSSelector` type should be added?
- [x] Documentation.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/3806)
<!-- Reviewable:end -->

[edit by @jgkamat to link issues: Closes #2773]

olmokramer commented on qutebrowser/config/configdata.yml:
```diff
@@ -954,6 +954,25 @@ hints.scatter:
+    inputs: []
+  type:
+    name: Dict
+    fixed_keys: ['all', 'links', 'images', 'urls', 'inputs']
```
I couldn't think of a way to use the `webelem.Group` enum's names here other than hardcoding it like this... Do you have any suggestions?

jgkamat commented on qutebrowser/browser/hints.py:
```diff
@@ -719,10 +719,21 @@ def start(self,  # pylint: disable=keyword-arg-before-vararg
+        selector = webelem.SELECTORS[group]
+
+        url = self._context.baseurl
+        group_name = webelem.Group(group).name
```
Can't this just be `group.name`?

olmokramer replied:
Yep, turns out it can be. I'm used to C enums so I figured `group` would just be an int...

olmokramer commented:
@jgkamat Thanks for the feedback!

1. I don't know if that would be useful at all. The selectors defined in `webelem.py` are so generic I can't really think of a reason to remove any of them. On the other hand, it is trivial to make them customisable, just define them in `configdata.yml` instead. Let's see what @The-Compiler thinks. A quick grep tells me there is just 1 other place in the codebase where `webelem.SELECTORS` is used (in `navigate.py`, to select links on the page), so that would also be easy to patch.
2. Would that mean that you could pass your own custom `--group` to the `:hint` command? That would be really cool! That would mean that if a the setting is set with a pattern matching the current URL but it doesn't have the specified group there has to be a fallback to the global setting.
3. Agreed, but I would prefer to just make the entire thing customisable because it's so damn easy.

jgkamat commented:
> Would that mean that you could pass your own custom --group to the :hint command?

Yup, that's exactly what I'm thinking! If custom groups were possible, you could add your own `:hint media` for example, to hint media elements, or similar. That example comes from #2773 (which I linked in the OP just now as well), but it opens up a lot of other cool customizability as well.

I didn't think about that edge case of per-domain settings, maybe we could just throw an error if we get undefined groups always, and have people bind to `[]` (which dosen't hint anything, I guess) if they want to squash the error? :P That case does need a bit more thought though.

olmokramer commented:
Thanks!

> I would really like a couple tests for this, maybe clicking divs and other elements with js loggers and custom hint classes, and making sure everything works well. Let me know if you'd like any help with that :)

Yeah, as I said in the first post I don't really know where to start with the tests, and what kind of tests to write for this. Some pointers would be nice :) I'm a bit swamped with study and work right now, though, so it may take me some time to do it.

> Also maybe @The-Compiler should take a high level look at the changes to sanity check it, and see if he likes the overall direction.

Yeah that's also part of the reason I haven't started writing tests yet, wanted to know if this was in general heading in the right direction before committing more time to it.

> It looks like moving the classes broke test_webkitelem, that dosen't look very fun to solve because of all the references... I'm not sure if this works as expected on qtwebkit, it's a bit tough for me to try it right now...

How would you suggest to fix those tests? I tried to use `config.val.hints.selectors` but `config.val` is `None` in the testfile. Anything special I must do to have access to the config in the tests?

olmokramer commented:
By the way, there's the `images` group now that selects only `img` elements, but I thought a more general `media` group that selects `audio, img, video` would be more useful. What are your thoughts? We probably shouldn't remove the `images` group though, as to not break backwards compatibility.

jgkamat commented:
Yup, there is a script which is supposed to do that, I think it's this one: https://github.com/qutebrowser/qutebrowser/blob/master/scripts/dev/src2asciidoc.py. Thanks again :)

olmokramer commented:
Ah didn't notice the `scripts/dev` directory... Much better this way :D Thanks!

olmokramer commented:
@mschilli87 Oops! :flushed: At least I learned a valuable lesson to read file headers :wink:

olmokramer commented:
@The-Compiler No worries, take your time :). I'm also swamped with other stuff.

That's a good point you raise about the default selectors, hadn't thought about that. What I probably would do in my `config.py` when I want to change the selectors for a specific domain/pattern is to copy the default and modify that, like below. That way I'll always use the default selectors.

```py
domain_all_sel = [*c.hints.selectors['all'], '.clickable']
config.set('hints.selectors', domain_all_sel, '*://domain/*')
```

olmokramer commented:
Ok I've added 2 basic tests, but one of them fails and I don't know how to fix it. It's the new test in `tests/end2end/features/hints/feature` on line 177 that checks if an error is shown when `:hint` is executed with an invalid hinting group. It seems to crash on line 179 (`And I hint with args "INVALID_GROUP"`) because an error is thrown on that line, before I can check if the correct error is shown...

The-Compiler commented on tests/end2end/features/hints.feature:
```diff
@@ -174,6 +174,11 @@ Feature: Using hints
 
+    Scenario: Error with invalid hint group
+        When I open data/hints/buttons.html
+        And I hint with args "INVALID_GROUP"
```
Using "And I hint with args ..." automatically waits for the hints to appear before continuing. You want `And I run :hint INVALID_GROUP` here instead.

olmokramer commented:
Hmm rebasing & squashing some commits seems to have messed up the commit history... Anyway, I've fixed that one test, thanks for the pointer!

I also updated the first post with more specific tests, please feel free to extend the list with things you feel should be tested for. I've also added the concern that users might set invalid CSS selectors, which we should probably test for. What are your thoughts?

jgkamat commented on qutebrowser/config/configdata.yml:
```diff
@@ -954,6 +954,69 @@ hints.scatter:
+      none_ok: true
+      valtype: String
+  supports_pattern: true
+  desc: CSS selectors used to determine which elements on a page should have
```
I think that the best way to solve the "users might override hints.selectors and miss out on updates" issue is to add documentation discouraging people from setting these values in autoconfig.yml. We could disable configuring this variable with `:set` too altogether as well.

Either way, I think this setting needs more description here than most, because it's very easy to mess up configuring this.

olmokramer commented:
@jgkamat while I mentioned validation of the user-input value a while ago, I see that I forgot to mention that, as it is now, qutebrowser crashes when the selector is invalid and the user tries to hint. So that definitely should also be looked at, either by validating the user's value (which I like most) or by try-catching the exception raised by the invalid selector and showing a message to the user.

I've thought of a few solutions to the "users might override & miss out on updates" problem as well, but they all have their drawbacks.. They are, in order of most to least flexibility:

1. Same as your suggestion, just document and tell people to be careful with this setting. Maybe add a small code snippet on how to *extend* rather than *replace* the defaults. Still easy to mess up, though.
2. If we're going to implement validation for the CSS selectors, we might as well check if the built-in hint groups are present in the user value, and if not add them anyway and/or show a warning. Still easy to override/remove an entire group, but other groups keep the default value.
3. Have fixed default groups, and only let users extend them. No option to remove unwanted selectors, though.

In any case I think adding a section to the FAQ or the qute://help/configuring.html with an example of how to properly and safely apply this setting would be useful.

olmokramer commented:
PS: What about the test cases I mentioned in OP but aren't checked yet?

slackhead commented:
What would be great is if there's a way of somehow hinting/selecting those overflow-x scrollbars so we can scroll sideways in those code/text boxes.

So far it seems to require a mouse click. I've tried searching and highlighting some text in one, but it still wouldn't scroll sideways.

I guess that this would require a different approach though, unless hinting a text box would work.

jgkamat commented:
> Domain specific hints. How would I go about this? I could add an HTML file with a <div class="clickable"> and a JS click handler on that div, but what domain pattern can I use?

I haven't tried this yet, but you might be able to use per-domain settings on a specific filename, for example:

`set --pattern="*://*/detect/is-javascript-enabled" content.javascript.enabled true`

enables js on `https://www.whatismybrowser.com/detect/is-javascript-enabled`.

You could also do unit tests instead of end2end for this, but that would be less "complete" :P

The-Compiler commented:
I planned to take a look at this during my holidays before starting to learn for exams, but I didn't manage to do so (mainly because of a lot of issues with Qt 5.11...). I'll get back to it in September though!

FWIW I want to think about this some more. I'm not too happy about having the selectors fixed in the config once the user customizes something or adds her own selector. Maybe there should be a generic way to add/remove/change values in a dict in the config, without affecting the entire value? Right now we have something like that for `bindings.default` and `bindings.command`, but there, it's just the keybinding code which merges the two values.

olmokramer commented:
@The-Compiler Haven't had a lot of time myself to work on this anyway. Good
luck with your exams! Just finished my last one today :D

In config.py it is already possible to dynamically change the values, e.g.

``` python
c.hints.selectors['all'].remove('iframe')
c.aliases['help'] = 'help --tab'
del c.url.searchengines['ddg']
c.aliases.update({ 'help': 'help --tab' })
# etc.
```

I think it would be enough disallow overwriting values at specific keys in the
dict, e.g. `c.hints.selectors['all'] = []` would raise an error.  The "locked"
keys could be (a subset of) the ones defined in configdata.yml. This still
allows to clear the default lists entirely (albeit a bit less efficient than
simply replacing the list with an empty one) and is simple to implement.

The autoconfig.yml case is a bit more complicated, because user can't just
freely modify the config value. The dict set in autoconfig.yml must be merged
with the default dict, otherwise users won't have updated defaults. But merging
of dicts is not always straightforward because dicts themselves can contain
dicts or lists, which in turn can contain dicts or lists etc. Are these
sub-values also merged? Should it be specified for each case? Also, merging
makes it impossible to unset default values in autoconfig.yml. I have never
used autoconfig.yml myself, so I don't know if this is already the case.

@jgkamat, @slackhead Sorry I missed your replies earlier.

> I haven't tried this yet, but you might be able to use per-domain settings on
> a specific filename, for example: 
> `set --pattern="*://*/detect/is-javascript-enabled" content.javascript.enabled true`

@jgkamat That's clever! :)

@slackhead It is possible to do what you are asking with this PR, but it is
a bit of a hack. You could write a greasemonkey script that inspects (a subset
of) elements on the page and adds a class or attribute to all elements whose
contents overflow their size. E.g.

```js
if (el.scrollWidth > el.clientWidth and getComputedStyle(el).overflowX == 'auto') {
	el.classList.add('__qutebrowser_overflow__');
}
```

Note that more code is needed to iterate over all candidate elements, also ones
dynamically added to the page. This may hurt performance a bit.

Then in your config.py add

```python
c.hints.selectors['overflow'] = ['.__qutebrowser_overflow__']
```

Now you can hint these elements in qutebrowser with `:hint overflow`.

The-Compiler commented:
Thank you! Merged with a couple of changes on top:

- 121483aa900a106d54ebf57cfdefcc64ff3b976d adds error handling for invalid selectors - I don't see a way to validate them earlier in the config (at least not without a new dependency, or hacks like creating a webpage only for validation). With this, at least users get a good error message.
- 0423ec6f9113d320a372a24b83a41fff2be88dd6 adds `webelem.css_selector`, so `:navigate` also merges global/per-URL settings properly. Before that, `:navigate` crashes when the user removed the `links` selector.
- f36285658e8c60ec0ed0d96ce5cdb700629e7434 disallows setting `hints.selectors` in `autoconfig.yml`, which should help with the "hardcoded default" issue for now.
- 3fe64085f8b91b6d8ee6b17665e2e6ae7e8a1bdb adds `[tabindex]` to the default selectors.

olmokramer commented:
Cool! And thanks for taking over, would've probably taken me much longer to make those changes. So glad that this is in now!
